airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py

```diff
@@ -0,0 +1,165 @@
+
+    @staticmethod
+    def add_secret_to_env(env, secret):
+        env.append({
+            'name': secret.deploy_target,
+            'valueFrom': {
+                'secretKeyRef': {
+                    'name': secret.secret,
+                    'key': secret.key
+                }
+            }
+        })
+
+    @staticmethod
+    def extract_labels(pod, req):
+        req['metadata']['labels'] = req['metadata'].get('labels', {})
+        for k, v in six.iteritems(pod.labels):
+            req['metadata']['labels'][k] = v
+
+    @staticmethod
+    def extract_cmds(pod, req):
+        req['spec']['containers'][0]['command'] = pod.cmds
+
+    @staticmethod
+    def extract_args(pod, req):
+        req['spec']['containers'][0]['args'] = pod.args
+
+    @staticmethod
+    def extract_node_selector(pod, req):
+        if len(pod.node_selectors) > 0:
+            req['spec']['nodeSelector'] = pod.node_selectors
+
+    @staticmethod
+    def attach_volumes(pod, req):
+        req['spec']['volumes'] = pod.volumes
+
+    @staticmethod
+    def attach_volume_mounts(pod, req):
+        if len(pod.volume_mounts) > 0:
+            req['spec']['containers'][0]['volumeMounts'] = (
+                req['spec']['containers'][0].get('volumeMounts', []))
+            req['spec']['containers'][0]['volumeMounts'].extend(pod.volume_mounts)
+
+    @staticmethod
+    def extract_name(pod, req):
+        req['metadata']['name'] = pod.name
+
+    @staticmethod
+    def extract_volume_secrets(pod, req):
+        vol_secrets = [s for s in pod.secrets if s.deploy_type == 'volume']
+        if any(vol_secrets):
+            req['spec']['containers'][0]['volumeMounts'] = []
+            req['spec']['volumes'] = []
+        for idx, vol in enumerate(vol_secrets):
+            vol_id = 'secretvol' + str(idx)
+            req['spec']['containers'][0]['volumeMounts'].append({
+                'mountPath': vol.deploy_target,
+                'name': vol_id,
+                'readOnly': True
+            })
+            req['spec']['volumes'].append({
+                'name': vol_id,
+                'secret': {
+                    'secretName': vol.secret
+                }
+            })
+
+    @staticmethod
+    def extract_env_and_secrets(pod, req):
+        env_secrets = [s for s in pod.secrets if s.deploy_type == 'env']
+        if len(pod.envs) > 0 or len(env_secrets) > 0:
+            env = []
+            for k in pod.envs.keys():
+                env.append({'name': k, 'value': pod.envs[k]})
+            for secret in env_secrets:
+                KubernetesRequestFactory.add_secret_to_env(env, secret)
+            req['spec']['containers'][0]['env'] = env
+
+    @staticmethod
+    def extract_resources(pod, req):
+        if not pod.resources or pod.resources.is_empty_resource_request():
+            return
+
+        req['spec']['containers'][0]['resources'] = {}
+
+        if pod.resources.has_requests():
```

Comment:
Is it easy to support more arbitrary pod resources requests/limits? For example, 'nvidia.com/gpu'. Would be nice to be able to request GPU resources.

Reply:
@kevincvlam what would you recommend for the user interface for that? I'd want to avoid forcing users to create a dict.

Reply:
@kevincvlam 
There is definitely a tradeoff here between extensibility and type-safety/readability of the api. 

The simplest api would be to let user's write their pod as a nested dict that gets directly translated into kubernetes json, but that can be painful for some tasks.

Reply:
@dimberman @grantnicholas makes sense, and good points. I think @grantnicholas highlighted the trade-off well, and I'm not sure what approach is best. Might mull it over a bit more.

In the case that it's decided to avoid forcing users to create a dict, I'd request that 'nvidia.com/gpu' resource added in the "translation layer" :-).

Reply:
@kevincvlam Would you be willing to either PR this in once this PR is merged or leave a quick issue on https://github.com/bloomberg/airflow/issues? I think this is a good feature but It'll be a lot easier to add things in once the initial PR is merged.

Reply:
@dimberman Possibly.. for now created issue: https://github.com/bloomberg/airflow/issues/42. +1, this makes sense as a follow-up PR. Thanks!